netbench/netbench-collector/src/main.rs

```diff
@@ -18,24 +36,363 @@ pub struct Args {
+            loop {
+                let state: State = current_state
+                    .clone()
+                    .load(Ordering::Relaxed)
+                    .try_into()
+                    .expect("An invalid atomic u8 got constructed.");
+                if state == State::Finished {
+                    log_serving!(verbose, "We are done; stop serving state");
+                    break Err(io::Error::new(ErrorKind::Other, "Finished"));
+                }
+                let (mut socket, _) = match timeout(Duration::from_secs(5), listener.accept()).await
+                {
+                    Ok(Ok(o)) => o,
+                    _ => {
+                        log_serving!(verbose, "No peers made connection in the last 5 seconds; checking for finished status");
+                        continue;
+                    }
+                };
+                let served_state: State = current_state
+                    .load(Ordering::Relaxed)
+                    .try_into()
+                    .expect("An invalid atomic u8 got constructed.");
+                socket.write_all(&serde_json::to_vec(&StateMessage::from(served_state)).unwrap()).await?;
+                log_serving!(verbose, "Told peer we are in state {:?}", served_state);
+            }
+        }))
+    }
+}
+
+macro_rules! log_server {
+    ($verbose:ident, $($arg:tt)*) => {
+        if $verbose {
+            println!("[SERVER] {}", format!($($arg)*));
+        }
+    }
+}
+
+macro_rules! log_client {
+    ($verbose:ident, $($arg:tt)*) => {
+        if $verbose {
+            println!("[CLIENT] {}", format!($($arg)*));
+        }
+    }
+}
+
+fn server_state_machine(args: Args, mut state_tracker: StateTracker) -> JoinHandle<io::Result<()>> {
+    let verbose = state_tracker.verbose;
+    tokio::spawn(async move {
+        state_tracker.store(State::Ready);
+        log_server!(verbose, "Ready to go; waiting to Client to report Ready.");
+        join!(state_tracker.poll(
+            State::Ready,
+            State::NotReady,
+            Duration::from_secs(5),
+            Duration::from_secs(5)
+        ))
+        .0
+        .unwrap()
+        .unwrap();
+        log_server!(verbose, "Client is ready; Starting Server");
+        state_tracker.store(State::Running);
+        let (poll, child) = try_join!(
+            state_tracker.poll(
+                State::Finished,
+                State::Finished,
+                Duration::from_secs(1),
+                Duration::from_secs(5)
+            ),
+            run(args)
+        )?;
+        poll?;
+        log_server!(verbose, "Client is Finished; We need to kill the server.");
+        child.kill().expect("Failed to kill child?");
+        log_server!(verbose, "Finished");
+        state_tracker.store(State::Finished);
+        Err(io::Error::new(ErrorKind::Other, String::from("Finished")))
+    })
+}
+
+fn client_state_machine(args: Args, mut state_tracker: StateTracker) -> JoinHandle<io::Result<()>> {
```

Comment:
perhaps rather than having this yield a join handle we could make is async and just await it?

Reply:
I don't have to use a join handle here for sure. There are other places in the code where I'm still using them. My attempt to convert them all ran into some type errors; I'll try again tomorrow.